### ISSUE TITLE
3.5.0 - it.only no longer works changing while test is open when you have a global after()

### ISSUE THREAD

**Issue by lukeapage** [cypress-io/cypress#5441]
### Current behavior:

When I change one test to `it.only` I end up with:

![image](https://user-images.githubusercontent.com/309321/67485091-ad3d8580-f669-11e9-9600-f7af4e1872cc.png)

The error is here:

```js
Cypress.runner.onRunnableRun = function (runnableRun, runnable, args) {
  debug('_onRunnableRun')

  const r = runnable
  const isHook = r.type === 'hook'
  const isAfterAllHook = isHook && r.hookName.match(/after all/)
```

where `r.hookName` is `undefined`.

![image](https://user-images.githubusercontent.com/309321/67486727-d57ab380-f66c-11e9-80fe-9d1d61842626.png)

### Desired behavior:

Works as in 3.4.3

### Steps to reproduce: (app code and test code)

1. create a spec with 2 tests
2. use [skip and only ui plugin](https://github.com/bahmutov/cypress-skip-and-only-ui)
3. open cypress
4. open the spec
5. edit the file to change one test to it.only

Note: workaround, closing the window and re-opening will work

It seems to be caused by having a global `after`: https://github.com/bahmutov/cypress-skip-and-only-ui/blob/master/src/support.tsx#L61

so I'm raising here as it seems like its more likely a global cypress issue, but of course it could be skip and only ui

### Versions

windows, cypress 3.5.0, running in electron

**Comment by jennifer-shehane** [comment]
@lukeapage Your issue can be addressed more quickly if you can provide the exact test code for us to run to see the issue. 

Working on following your steps to recreate currently.

**Comment by jennifer-shehane** [comment]
This is not reproducible in Cypress without the https://github.com/bahmutov/cypress-skip-and-only-ui plugin. Please open an issue in that repo. The below code works. 

```js
context('Connectors', () => {
  it.only('test 1', () => {
    cy.visit('https://example.cypress.io/commands/connectors')
    cy.get('.connectors-each-ul>li')
  })

  it('test 2', () => {
    cy.visit('https://example.cypress.io/commands/connectors')
    cy.get('.connectors-its-ul>li')
  })

  after(() => {
    cy.log('after')
  })
})
```

**Comment by lukeapage** [comment]
this reproduces it:

https://github.com/cypress-io/cypress-test-tiny/pull/51

**Comment by lukeapage** [comment]
its not always straight away, but if you try changing `it.only` on the fly it happens after 1 or 2 tries

**Comment by lukeapage** [comment]
from that pr:

![image](https://user-images.githubusercontent.com/309321/67488836-c564d300-f670-11e9-8c7a-09922b826320.png)

**Comment by lukeapage** [comment]
it actually happens at the end of executing the test where it has a `it.only` - so if I change to it.only, save, then remove it.only and save, I am left in a broken state

**Comment by lukeapage** [comment]
@jennifer-shehane please re-open this

**Comment by jennifer-shehane** [comment]
I'm not able to reproduce this with the example provided either. 😞 

Did you isolate this to only happening in Electron? I am running on MacOS Mohave. What is the IDE you're using?

<img width="1361" alt="Screen Shot 2019-10-24 at 9 35 37 AM" src="https://user-images.githubusercontent.com/1271364/67490714-aa2e9f00-f641-11e9-8d0f-519a442a4472.png">

**Comment by jennifer-shehane** [comment]
I did get this console error once, the UI was still accurate though. It may be completely unrelated. 

<img width="999" alt="Screen Shot 2019-10-24 at 9 39 03 AM" src="https://user-images.githubusercontent.com/1271364/67491067-4c4e8700-f642-11e9-8696-b4a3cdf0e943.png">


Can you open the console and print any errors printed there please?

**Comment by lukeapage** [comment]
I use webstorm, but I reproduced it also with notepad++ (see screenshot).
Did you see this comment:
> it actually happens at the end of executing the test where it has a it.only - so if I change to it.only, save, then remove it.only and save, I am left in a broken state

Because your screenshot is after the `it.only` but not after going back to 2 tests.

It may be that this bug occurs when editing the test whilst it is running rather than after it has finished (so a wait may help?)

I am on windows, but by the looks of the js exception, I would be surprised if it was os specific.

I can reproduce in chrome and electron.

**Comment by lukeapage** [comment]
I get this in the chrome console
```
Uncaught Error: runnable must have an id
    at Object.onRunnableRun (cypress_runner.js:103453)
    at $Cypress.action (cypress_runner.js:97300)
    at Hook.Runnable.run (cypress_runner.js:102519)
    at next (cypress_runner.js:31274)
    at cypress_runner.js:31296
    at timeslice (cypress_runner.js:26364)
```

![image](https://user-images.githubusercontent.com/309321/67491382-268ea580-f675-11e9-959e-699ac7411fda.png)

And after I get that error, on the next edit of my text editor and save - the spec window is broken forever

**Comment by lukeapage** [comment]
yes the runnable without an id is the global `after` function

![image](https://user-images.githubusercontent.com/309321/67491599-7e2d1100-f675-11e9-848f-2f2d272ecda6.png)

**Comment by jennifer-shehane** [comment]
>it actually happens at the end of executing the test where it has a it.only - so if I change to it.only, save, then remove it.only and save, I am left in a broken state

Yah, I am doing this and can't recreate.

Do you have manual saving on in your IDE? Some IDEs auto-save as you type after some delay, just wondering if that is an aspect. I only have manual saving on.

Thanks for sending the console error, that is helpful.

**Comment by lukeapage** [comment]
> Do you have manual saving on in your IDE? Some IDEs auto-save as you type after some delay, just wondering if that is an aspect. I only have manual saving on.

Only manual saving and as i said, I tried 2 editors

**Comment by lukeapage** [comment]
I can try it on mac tomorrow to see if its windows specific

**Comment by bahmutov** [comment]
Hmm, could not recreate this on Mac. Tried adding skip ony ui plugin - but the plugin does not seem to work at all.

**Comment by lukeapage** [comment]
I can reproduce on mac too - but it was harder - I added wait commands and then switched between it and it.only 5 or 6 times whilst the command is waiting and then it breaks.

So it probably is not OS dependent but timing dependent. Maybe you're on faster macs than me or something along those lines.

This is again just using my PR to cypress-test-tiny.

![image](https://user-images.githubusercontent.com/309321/67553121-bd5a7100-f70c-11e9-92f8-4d3b241e1464.png)

**Comment by lukeapage** [comment]
Note that when it breaks the timer continues forever and No commands were issues during test is present under both tests. Editing the file again does not fix it.

This is relatively low priority because we can just stop using skip and only ui plugin, but if I were to need a global `after` function it would make development hell.

**Comment by lukeapage** [comment]
This is no longer reproducible in 4.2.0

**Comment by lukeapage** [comment]
Oops, this is reproducible, forgot to read all my steps - it doesn't happen straight away but only after a few times of editing the spec file.

**Comment by jennifer-shehane** [comment]
I'm still not able to reproduce this. I was suspecting that this PR in 4.2.0 should have fixed this. https://github.com/cypress-io/cypress/pull/6620

Honestly, there's nothing we can do without being able to reproduce this though.

**Comment by jennifer-shehane** [comment]
Right now there doesn't seem to be enough information to reproduce the problem on our end. We'll have to close this issue until we can reproduce it. This does not mean that your issue is not happening - it just means that we do not have a path to move forward. 

Please comment in this issue with a reproducible example and we will consider reopening the issue. Here are some tips for providing a [Short, Self Contained, Correct, Example](http://sscce.org/) and our own [Troubleshooting Cypress](https://on.cypress.io/troubleshooting) guide.
